core/services/workflows/engine.go

```diff
@@ -30,6 +30,52 @@ type stepRequest struct {
 	state   store.WorkflowExecution
 }
 
+type stepUpdateChannel struct {
+	executionID string
+	ch          chan store.WorkflowExecutionStep
+}
+
+type stepUpdateManager struct {
+	mu sync.RWMutex
+	m  map[string]stepUpdateChannel
+}
+
+func (sucm *stepUpdateManager) add(executionID string, ch stepUpdateChannel) (added bool) {
+	sucm.mu.Lock()
+	defer sucm.mu.Unlock()
+	if _, ok := sucm.m[executionID]; ok {
+		return false
+	}
+	sucm.m[executionID] = ch
+	return true
+}
+
+func (sucm *stepUpdateManager) remove(executionID string) {
+	sucm.mu.Lock()
+	defer sucm.mu.Unlock()
+	if _, ok := sucm.m[executionID]; ok {
+		close(sucm.m[executionID].ch)
+		delete(sucm.m, executionID)
+	}
+}
+
+func (sucm *stepUpdateManager) send(ctx context.Context, executionID string, stepUpdate store.WorkflowExecutionStep) error {
+	sucm.mu.RLock()
+	stepUpdateCh, ok := sucm.m[executionID]
+	if !ok {
+		sucm.mu.RUnlock()
+		return fmt.Errorf("step update channel not found for execution %s, dropping step update", executionID)
+	}
+	sucm.mu.RUnlock()
+
+	select {
+	case <-ctx.Done():
+		return errors.New("context canceled before step update could be issued")
```

Comment:
You could wrap the `ctx.Err()` or even better `context.Cause(ctx)`:
```suggestion
		return errors.New("context canceled before step update could be issued: %w", context.Cause(ctx))
```